pilot/pkg/networking/core/route/route.go

```diff
@@ -408,16 +409,25 @@ func BuildHTTPRoutesForVirtualService(
 	out := make([]*route.Route, 0, len(vs.Http))
 
 	catchall := false
+	var lastMatch *route.RouteMatch
 	for _, http := range vs.Http {
 		if len(http.Match) == 0 {
 			if r := TranslateRoute(node, http, nil, listenPort, virtualService, gatewayNames, opts); r != nil {
+				if lastMatch != nil && proto.Equal(lastMatch, r.Match) {
+					continue
+				}
 				out = append(out, r)
+				lastMatch = r.Match
 			}
 			catchall = true
 		} else {
 			for _, match := range http.Match {
 				if r := TranslateRoute(node, http, match, listenPort, virtualService, gatewayNames, opts); r != nil {
+					if lastMatch != nil && proto.Equal(lastMatch, r.Match) {
+						continue
+					}
 					out = append(out, r)
+					lastMatch = r.Match
```

Comment:
Isn't this only comparing adjacent duplicates...? Like would `[a, b, a]` match this?